### ISSUE TITLE
Use as Generic Text-based Collaboration Backend

### ISSUE THREAD

**Issue by ziadsawalha** [codimd/server#494]
We'd love to use CodiMD to also collaborate on other text-based documents such as [OpenAPI (a.k.a Swagger)](https://editor.swagger.io/), [AsyncAPI](http://editor.asyncapi.org/), and [JSON Schema](https://jsonschema.net/home) documents.

We might contribute some code for this, but wanted your take on it first.

- Have you considered usage of the backend for anything other than markdown?
- Do you foresee any issues with using the 2.0 backend in that way?
- Would you consider a feature to allow the front-end to toggle between formats?

**Comment by DerMolly** [comment]
Generally speaking this should be possible, but there are some things I want to point out:

> Have you considered usage of the backend for anything other than markdown?

We currently are in the process of rewriting the whole thing to make it more flexible, and easier to understand for new developers. This should hopefully bring more features and better user experience in the long run. This also means priority number 1 is getting most of the features working again and only some improvements. As you can imagine a whole other document type was not on the shortlist of things to do.

> Do you foresee any issues with using the 2.0 backend in that way?

If I recall the plans for the 2.0 backend correctly, the only thing that may be a hindrance, could be the way we handle opengraph integrations. But as the functionality for that should reside in the frontend, annother frontend should be able to work around that. That said I don't think we currently plan to work on a second frontend as our dev-resources are rather thin and this seems to be very extra (but cool nonetheless).

> Would you consider a feature to allow the front-end to toggle between formats?

This maybe a way to go, but I'd consider much further down the road. I'm not even sure this would be something I'd consider for a 2.x release.

That being said I guess a good compromise could be to add OpenAPI (a.k.a Swagger), AsyncAPI, and JSON Schema as languages to highlight.js (that's our highlighter for the renderer). That way you could work on these documents as code blocks in HedgeDoc. Instead (or additionally) one could add a special rendering of these kind of code blocks as we currently do with [mermaid](https://demo.codimd.org/#Mermaid) for example (at least for OpenAPI there exists this rather nice interactive way to view the api spec). 

TL;DR
I don't think we will add another frontend or document type to our existing frontend any time soon, but maybe some support for OpenAPI (a.k.a Swagger), AsyncAPI, and JSON Schema could be added in the form of OpenAPI (a.k.a Swagger), AsyncAPI, and JSON Schemacode block

**Comment by ziadsawalha** [comment]
Thank you for the prompt and detailed response, @DerMolly 
> As you can imagine a whole other document type was not on the shortlist of things to do.

Does the backend actually know the content is markdown? Or does it just treat it as a string?

**Comment by DerMolly** [comment]
It handles it mostly as a string, but there is the possibility to set the content of a note via api. The endpoint expects a file of the mime type text/markdown.

**Comment by SISheogorath** [comment]
One should also note, that we had similar debates for different rendering engines before and back then we always stepped back from adding additional complexity, as in another renderer, from the project. We are definitely focused on markdown and taking this assumption away from the project would probably make the project scope exploding due to the fact how generic one would need to build things all of the sudden.

I highly recommend against going down that road as it makes everyone's life harder. I may repeat myself but building things as specific as possible and only as generic as needed makes sure that we stay focused and build the best tools for the problem we try to solve (which currently is writing Markdown collaboratively).

But probably @davidmehren is the best person to address this conversation as he's the technical lead for the backend code and might has different thoughts on it :)

**Comment by ccoenen** [comment]
For "our" frontend/client I agree completely. Even "markdown only" is enough work as it is. That said: the backend (in version 1.6) is operational transform on top of Socket.io backed by a database. Whatever is piped in there is pretty much up to you. Theoretically it should even be OK with binary data (though I can't speak from experience and do not recommend it).

anyway: if you build your own client on top of that, you'll probably be fine. But you will also be on your own. Please note that the server is also changing for version 2.0, but I am not working on it, so I can't grok how fundamental those changes will be.

**Comment by ziadsawalha** [comment]
Sounds like writing our own client is preferable to complicating the current one. We're fine with that and - if we do it - we will keep an eye out on how much it diverges.

> operational transform on top of Socket.io backed by a database

This is a really valuable feature. @davidmehren - do you anticipate functionality coming to the backend in 2.0 that would make it hard to support content other than markdown? Even the "endpoint expects a file of the mime type text/markdown" sounds safe to extend by supporting other content types.

Thank you @ccoenen @SISheogorath @DerMolly for your responses.

**Comment by ccoenen** [comment]
You could just ignore the mimetype at your end. ¯\_(ツ)_/¯

Here's some documentation on the OT part:
https://github.com/codimd/server/blob/master/docs/dev/ot.md

I'm not sure if a generic text backend is something we should do, in general. But for 2.0 I would be especially against it, because it is not helpful to widen the scope even more at this point.

**Comment by InnayTool** [comment]
I think as @ccoenen discribed for the 1.6 backend it should work to use the 2.0 backend for other text formats. We try to avoid dependencies to markdown because it's not needed for almost all features, so it should be possible to build your own frontend on top of that. A problem could be API features and that we don't plan a stable API with versions for communication with the frontend. @ziadsawalha if you have input for the backend feel free to join the discussion in the issues and the [development chat](https://matrix.to/#/#hedgedoc-dev:matrix.org).

**Comment by davidmehren** [comment]
A similar question came up recently with https://github.com/hedgedoc/hedgedoc/issues/1008. Have a look over there for more details about the current state regarding extensibility of HedgeDoc.

I plan to keep this issue open to track possible refactoring work.
